### problem solving and programming
##### 12th june 2019
###### day objectives

#### string slicing

In [ ]:
a1="saikrishna"
a1[0]
a1[len(a1)-1]
a1[-2]
a1[1:-1]
a1[len(a1)//2]
a1[-1::-1]
a1[-1:-3:-1]

a1[:2]

#### functions

In [ ]:
def  plindrome(s):
    if s==s[::-1]:
        return True
    else:
        return False
plindrome('123321')

In [ ]:
def checkleapyear(year):
    if (year % 400 ==0 ) or (year %100!=0) and (year %4==0):
        return True
    return False
checkleapyear(2020)

In [ ]:
#count the nunber of digits in number
def digi(h):
    k=len(str(h))
    return k
digi(1234556)
    

In [ ]:
#greatest of numbers
def great():
    a=int(input())
    b=int(input())
    c=int(input())
    d=int(input())
    p=[]
    p.append(a)
    p.append(b)
    p.append(c)
    p.append(d)
    l=max(p)
    return(l)
great()
    

### ITERATION
* FOR
* WHILE

In [11]:
#N NATURAL NUMBERS
def printnaturalnumbers(n):
    for i in range(1,n+1):
        print(i,end=" ")
    return
 
printnaturalnumbers(10)

1 2 3 4 5 6 7 8 9 10 

In [ ]:
def printnaturalnumbers(n):
    c=1
    while c<=n:
        print(c)
        c=c+1
    return

### Problem Set
* Function to print all numbers divisible by 6 and not a factor of 100 in a given range(lb, ub) inclusive
* Function to find the average of cubes of all even numbers in a given range(lb, ub) inclusive
* Function to generate the list of factors for a given number
* Function to calculate the factorial of a given number
* Function to check if a given number is Prime
* Function to calculate the average first N Prime numbers
* Function to generate all Perfect numbers in a given range

In [40]:
#Function to print all numbers divisible by 6 and not a factor of 100 in a given range(lb, ub) inclusive

def printallnumbers(l,u):
   for i in range(l,u+1):
       if (i % 6==0) and (100 % i!=0):
          print(i, end=" ")
   return
printallnumbers(100,200)

102 108 114 120 126 132 138 144 150 156 162 168 174 180 186 192 198 

In [ ]:
#Function to find the average of cubes of all even numbers in a given range(lb, ub) inclusive
def printcubesnumbers(lb,ub):
   count=0
   sum=0
   for i in range(lb,ub+1):
       if i%2==0:
           #print(i)
           k=i**3
           #print(k)
           count=count+1
           sum=sum+k
           #print(k)
           #print(sum)
           #print(count)
   p=sum/count
   return p
printcubesnumbers(1,10)

In [ ]:
#Function to generate the list of factors for a given number
def printfactors(n):
   for i in range(1,n+1):
       if n%i==0:
           print(i)
   return
printfactors(122)

In [41]:
#Function to calculate the factorial of a given number
def fact():
    s=1
    l=int(input('enter number : '))
    for i in range(1,l+1):
        s=s*i
    return s
fact()

enter 1st no  : 5


120

In [44]:
#Function to check if a given number is Prime



enter 1st no  : 4


True

In [46]:
#Function to calculate the average first N Prime numbers
def primeaverage():
    
    l=int(input('enter 1st no  : '))
    u=int(input('enter 2nd no  : '))
    sum=0
    count=0
    for num in range(l,u + 1):
        if num > 1:
           for i in range(2,num):
              #print(i)
              if (num % i) == 0:
                   break
           else:
            #print(num)
            sum=sum+num
            count=count+1
    return (sum/count)
    
primeaverage()

enter 1st no  : 1
enter 2nd no  : 10


4.25

In [47]:
#Function to generate all Perfect numbers in a given range

def factors_sum(i):
   sum=0
   for j in range(1,i):
       if i%j==0:
           sum+=j
   return sum
def perfect(lb,ub):
   sum=0
   for i in range(lb,ub):
       x=factors_sum(i)
       if i==x:
           print(i,end=" ")
   return
perfect(1,1500)

6 28 496 

##### Advanced Problem Set( Optional )
* Function to calculate average of all factorials in a given range
* Function to generate N odd armstrong numbers 
* Function to generate Multiplication table for a number in a given range
    10 in the range(100, 102) inclusive
    10 x 100 = 1000
    10 x 101 = 1010
    10 x 102 = 1020

In [39]:
#Function to calculate average of all factorials in a given range
def factaverage():
    s=1
    sum=0
    l=int(input('enter 1st no  : '))
    u=int(input('enter 2nd no  : '))
    for i in range(1,u+1):
        s=s*i
        #print(s)
        sum=sum+s
    return sum/u
factaverage()

enter 1st no  : 1
enter 2nd no  : 5


30.6

In [52]:
#Function to generate N odd armstrong numbers
def oddarmstrong():
    l=int(input('enter 1st no  : '))
    u=int(input('enter 2nd no  : '))
    sum=0
    
    for i in range(l,u+1):
        t=i
        sum=0
        o=len(str(i))
        while t<0:
            k=t%10
            sum=sum+K**o
            t//=10
    if i==sum:
        print (sum)
    return sum

oddarmstrong()
            
    
    
    

enter 1st no  : 1
enter 2nd no  : 100


0